src/Akeneo/Bundle/ElasticsearchBundle/Cursor/AbstractCursor.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Akeneo\Bundle\ElasticsearchBundle\Cursor;
+
+use Akeneo\Bundle\ElasticsearchBundle\Client;
+use Akeneo\Component\StorageUtils\Cursor\CursorInterface;
+use Akeneo\Component\StorageUtils\Repository\CursorableRepositoryInterface;
+
+/**
+ * Common logic shared by all our cursors.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+abstract class AbstractCursor implements CursorInterface
+{
+    /** @var Client */
+    protected $esClient;
+
+    /** @var CursorableRepositoryInterface */
+    protected $repository;
+
+    /** @var array */
+    protected $esQuery;
+
+    /** @var string */
+    protected $indexType;
+
+    /** @var array */
+    protected $items;
+
+    /** @var int */
+    protected $pageSize;
+
+    /** @var int */
+    protected $count;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function current()
+    {
+        if (null === $this->items) {
+            $this->rewind();
+        }
+
+        return current($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function key()
+    {
+        if (null === $this->items) {
+            $this->rewind();
+        }
+
+        return key($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function valid()
+    {
+        if (null === $this->items) {
+            $this->rewind();
+        }
+
+        return !empty($this->items);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function count()
+    {
+        if (null === $this->items) {
+            $this->rewind();
+        }
+
+        return $this->count;
+    }
+
+    /**
+     * Get the next items (hydrated from doctrine repository).
+     *
+     * @param array $esQuery
+     *
+     * @return array
+     */
+    protected function getNextItems(array $esQuery)
+    {
+        $identifiers = $this->getNextIdentifiers($esQuery);
+        if (empty($identifiers)) {
+            return [];
+        }
+
+        $hydratedItems = $this->repository->getItemsFromIdentifiers($identifiers);
```

Comment:
because there is no _easy_ way to do that in MySQL

- you retrieve the identifiers ordered from ES
- then you load them from MySQL (impossible to keep the order here easily with Doctrine)
- then you have to re order them in the way ES returned them

basically, we'd like to do that https://stackoverflow.com/questions/958627/mysql-order-by-values-within-in